src/app/Library/CrudPanel/Traits/Create.php

```diff
@@ -2,7 +2,7 @@
 
```

Comment:
Food for thought:

Now that I think about it, I'd feel A LOT more confident about this if we had tests for these use cases. It feels like something important but also super-difficult to debug, so... exactly the type of thing that should be unit tested.

The problem is though - WHAT would we test? 
- Would we test just the new methods for relationships? That'd be a good idea anyway, but it wouldn't be a comprehensive test that confirms a use case is working fine. 
- Would we test the  `create()` and `update()` methods, when the `relationship` column is used? That would be the comprehensive use case test above, but... wouldn't that become unit tests for the `relationship` column then?
- Would we create HTTP/feature tests, using Laravel Dusk, like I started doing in the Demo? But then, the tests are basically in a different repo, which... I don't think is a good idea. Cause the tests won't be automatically run when we push changes, so we'll still be able to push breaking changes. So it defeats the purpose.

Which brings the question. Which we should think about, for now or for v12 when we rewrite fields to objects... **How will we write tests for field types?** Because if fields had tests... man... it'd be SO MUCH easier to be confident we're not breaking anything in our changes, and to discover broken/uncovered use cases.

I have no practical solution for this. I'm just wondering out loud, for you guys to also think about this. How can we become more confident that a field works as expected, for all the use cases it covers? And how do we do that in a way that even third-party vendors, who create Backpack fields, can write them the same way, pretty easily, to make sure all their use cases are covered? Or am I thinking about this the wrong way?

Reply:
Rephrased - what should we write tests for? The operation? Or the fields/columns that the operation uses?

Reply:
I 100% agree with you. I'v been having a look at this. I think to properly cover the full process we need a mix of browser tests and the unit tests. 

We already have some coverage to the relationships, belongs to is missing but I am adding it. Thing is in this tests we are providing the data (field names and formats) in a way that is expected by the controller to receive from browser. 

At the moment it basically tests the whole process, but we are MITM between controller show form and submit form.

To cover that we need some browser tests too:

```php
// create some crud panel with some fields

// navigate to panel url. assert what fields you get, what value they get or populate the form.

//hit submit wait for response and validate response data.
```
This will be specialy usefull to test field `configurations` like `ajax => true`, etc etc

In the meanwhile, testing the core and seeing the green is the best we can do, and I agree that we should do that.

I'v just pushed tests for belongs to relation and has one.